docs/compilers/CSharp/Runtime Async Design.md

```diff
@@ -0,0 +1,456 @@
+# Runtime Async Design
+
+See also the ECMA-335 specification change for this feature: https://github.com/dotnet/runtime/pull/104063, https://github.com/dotnet/runtime/blob/main/docs/design/specs/runtime-async.md (when the PR completes). https://github.com/dotnet/runtime/issues/109632 tracks open issues for the feature in the runtime.
+
+This document goes over the general design of how Roslyn works with the Runtime Async feature to produce IL. In general, we try to avoid exposing this feature at the user level; initial binding is almost entirely
+unaffected by runtime async. Exposed symbols do not give direct information about whether they were compiled with runtime async, and indeed the compiler has no idea whether a method from a referenced assembly is
+compiled with runtime async or not.
+
+## Supporting runtime apis
+
+We use the following runtime flag to drive whether feature can be used. This flag must be defined in the same assembly that defines `object`, and the assembly cannot reference any other assemblies. In terms of
+CoreFX, this means it must be defined in the `System.Runtime` reference assembly.
+
+TODO: Determine whether just the presence of this flag will cause the compiler to generate in runtime async mode.
```

Comment:
My 2 cents: I think the driving scenario here is going to be multi-targeting. Consider for example if I have an application that targets `net9.0` and `net11.0`. The build needs to be functional when having this multi-target and an explicit langversion of 14.0. Ideally for the `net9.0` target async v1 code generation occurs and for `net11.0` we get async v2 generation. 

Given that I believe that we should gate async v2 code generation on effectively

```csharp
bool IsAsyncV2 => LangVersion >= 14 && IsRuntimeFeatureFlagPresent(Async);
```


Note: my comment here specifically uses `net11.0` because that is where the feature flag will be present in the standard core lib. I would expect the experience with `net10.0` + use the async2 preview build flag matches `net11.0` 

Note2: I suspect that in the short term we will want a feature flag to explicitly use async v1 code generation. Basically override the defaults. Having that will allow us to quickly toggle between async v1 and v2 generation. Suspect that will be useful for quickly ruling out / in whether a behavior is tied to the new model or not.

Reply:
The async v2 model is going to be a _highly compatible_ design but not a _fully compatible_ one. There will be a few rough edges that will some segment of customers up. From that angle I can see this as being a language version issue because it allows a customer to separate out movement to a new runtime vs. a new language version of `async`. 

Irrespective of whether it's a language version feature, I do think we are going to want a feature flag that allows us to tweak the new codegen off. Particularly during `net10.0` and early `net11.0` it will be an effective way to help diagnose issues.

Reply:
Regardless of our goals I also suspect there will be bugs that will push for a flag to turn things on/off. I think an attribute like `MethodImplOptions` at the method level might even be appropriate, although I haven't given it much thought.

Reply:
> I think an attribute like MethodImplOptions at the method level might even be appropriate

Yes, I am assuming we will have this, and it's included in this document.

Reply:
> I do think we are going to want a feature flag that allows us to tweak the new codegen off. Particularly during net10.0 and early net11.0 it will be an effective way to help diagnose issues.

Oh, definitely. I figure for "turn off for the entire assembly", it'll be a `Features` element, and for individual methods, we'll use the attribute I defined in this doc.